test/test_utils.py

```diff
@@ -2060,22 +2060,40 @@ def test_extract_basic_auth(self):
 
     @unittest.skipUnless(compat_os_name == 'nt', 'Only relevant on Windows')
     def test_Popen_windows_escaping(self):
+        tests = [
+            'test"&',
+            '%CMDCMDLINE:~-1%&',
+            'a\nb',
+            '"',
+            '\\',
+            '!',
+            '^!',
+            'a \\ b',
+            'a \\" b',
+            'a \\ b\\',
+            # We replace \r with \n
+            ('a\r\ra', 'a\n\na'),
+        ]
+
         def run_shell(args):
             stdout, stderr, error = Popen.run(
                 args, text=True, shell=True, stdout=subprocess.PIPE, stderr=subprocess.PIPE)
             assert not stderr
             assert not error
             return stdout
 
-        # Test escaping
-        assert run_shell(['echo', 'test"&']) == '"test""&"\n'
-        assert run_shell(['echo', '%CMDCMDLINE:~-1%&']) == '"%CMDCMDLINE:~-1%&"\n'
-        assert run_shell(['echo', 'a\nb']) == '"a"\n"b"\n'
-        assert run_shell(['echo', '"']) == '""""\n'
-        assert run_shell(['echo', '\\']) == '\\\n'
-        # Test if delayed expansion is disabled
-        assert run_shell(['echo', '^!']) == '"^!"\n'
-        assert run_shell('echo "^!"') == '"^!"\n'
+        for argument in tests:
+            if isinstance(argument, tuple):
+                argument, expected = argument
+            else:
+                expected = argument
+
+            args = [sys.executable, '-c', 'import sys; print(end=sys.argv[1])', argument]
+            assert run_shell(args) == expected
```

Comment:
Instead of opening a subprocess each time, we could process all the tests in one go, right?

Reply:
While that would work, spawning a new process is not as expensive as I thought (Whole test takes about 1 second on my machine), and it would complicate test implementation

Reply:
Actually, we have to do something like
```suggestion
            args = [sys.executable, '-c', 'import sys; print(end=sys.argv[1])', argument, 'end']
            assert run_shell(args) == expected
```

Argument parsing is tolerant of missing `"` at end. So without another argument, we can't say for sure that the quoting was correct even if round-trip succeeded.

Reply:
Implementation to test everything at once:
```py
        arguments, expected = itertools.zip_longest(*map(variadic, tests))
        expected = [expect or arg for expect, arg in zip(expected, arguments)]
        expected.append('fail')

        args = [sys.executable, '-c', 'import sys; print(sys.argv[1:])', *arguments, 'fail']
        assert ast.literal_eval(run_shell(args)) == expected

        escaped = shell_quote(arguments, shell=True)
        args = f'{sys.executable} -c "import sys; print(sys.argv[1:])" {escaped} fail'
        assert ast.literal_eval(run_shell(args)) == expected
```
The downside to this is, if escaping fails for one item earlier in the list, this might make it hard to pinpoint exectly where the error is.

Reply:
I went with leaving it the way it is. Its better to leave it to explicit tests and if someone complains about *tests* being slow, we can always change it later. I did add an `end` to test proper quote balancing

Reply:
I like https://github.com/yt-dlp/yt-dlp/pull/9802#discussion_r1581754335, but not an issue